hashes/src/lib.rs

```diff
@@ -125,101 +119,109 @@ pub mod siphash24;
-    /// to implement the `io::Write` trait, and to never return errors under
-    /// any conditions.
-    type Engine: HashEngine;
-
-    /// The byte array that represents the hash internally.
-    type Bytes: hex::FromHex + Copy;
-
-    /// Constructs a new engine.
-    fn engine() -> Self::Engine { Self::Engine::default() }
-
-    /// Produces a hash from the current state of a given engine.
-    fn from_engine(e: Self::Engine) -> Self;
-
-    /// Length of the hash, in bytes.
-    const LEN: usize;
-
-    /// Copies a byte slice into a hash object.
-    fn from_slice(sl: &[u8]) -> Result<Self, FromSliceError>;
-
-    /// Hashes some bytes.
-    fn hash(data: &[u8]) -> Self {
-        let mut engine = Self::engine();
-        engine.input(data);
-        Self::from_engine(engine)
+    /// Returns the final digest from the current state of the hash engine.
+    fn finalize(self) -> Self::Digest;
+
+    /// Creates a default hash engine, adds `bytes` to it, then finalizes the engine.
+    ///
+    /// # Returns
+    ///
+    /// The digest created by hashing `bytes` with engine's hashing algorithm.
+    fn hash(bytes: &[u8]) -> Self::Digest {
+        let mut engine = Self::new();
+        engine.input(bytes);
+        engine.finalize()
     }
 
     /// Hashes all the byte slices retrieved from the iterator together.
-    fn hash_byte_chunks<B, I>(byte_slices: I) -> Self
+    fn hash_byte_chunks<B, I>(byte_slices: I) -> Self::Digest
     where
         B: AsRef<[u8]>,
         I: IntoIterator<Item = B>,
     {
-        let mut engine = Self::engine();
+        let mut engine = Self::new();
         for slice in byte_slices {
             engine.input(slice.as_ref());
         }
-        Self::from_engine(engine)
+        engine.finalize()
     }
 
-    /// Flag indicating whether user-visible serializations of this hash
-    /// should be backward. For some reason Satoshi decided this should be
-    /// true for `Sha256dHash`, so here we are.
-    const DISPLAY_BACKWARD: bool = false;
-
-    /// Returns the underlying byte array.
-    fn to_byte_array(self) -> Self::Bytes;
-
-    /// Returns a reference to the underlying byte array.
-    fn as_byte_array(&self) -> &Self::Bytes;
-
-    /// Constructs a hash from the underlying byte array.
-    fn from_byte_array(bytes: Self::Bytes) -> Self;
+    /// Outputs the midstate of the hash engine. This function should not be
+    /// used directly unless you really know what you're doing.
+    fn midstate(&self) -> Self::Midstate;
 
-    /// Returns an all zero hash.
+    /// Create a new [`HashEngine`] from a [`Self::Midstate`].
     ///
-    /// An all zeros hash is a made up construct because there is not a known input that can create
-    /// it, however it is used in various places in Bitcoin e.g., the Bitcoin genesis block's
-    /// previous blockhash and the coinbase transaction's outpoint txid.
-    fn all_zeros() -> Self;
+    /// Only use this function if you know what you are doing.
```

Comment:
I really don't like this handwavy documentation. Sometimes people think they know what they are doing but they are actually missing something.

Reply:
What would you suggest?

I am a professional cryptographer and maintainer of rust-bitcoin and I can assure you that I need this function available :P.